docs/collaborative-pinsets-rfc.md

```diff
@@ -0,0 +1,179 @@
+    RmPeer(p peer.ID) error
+    State() (state.State, error)
+    Leader() (peer.ID, error)
+    WaitForSync() error
+    Clean() error
+    Peers() ([]peer.ID, error)
+}
+```
+
+This is essentially a wrapper of the go-libp2p-consensus which adds functionality and utility methods which until now applied to raft (provided by go-libp2p-raft).
+
+The purpose of Raft is to maintain a shared state by providing a distributed append-only log. In a Raft cluster, the log is maintained by an elected Leader, compacted and snapshotted convieniently. IPFS Cluster has spent significant efforts in detaching the state representation from raft, and allowing transformations (upgrades) to run indepedently, based solely on the "state component".
+
+### A new consensus component using a IPFS-backed CRDTs
+
+In order to be consistent with how components are meant to interact with each-other, it makes sense to implement the shared state maintenance in a new `go-libp2p-consensus` implementation. The main pain points in a collaborative cluster will be:
+
+* Security: all peers should not be able to control or influence the behaviour of all other peers (including updating the state). For this, we will introduce the notion of a *trusted peerset* along with RPC endpoint authorization.
+* Scalability: this consensus layer should support very large number of peers, efficient broadcast of state updates. For this we will introduce a CRDT-based "consensus" layer (it's not consensus per-se) and rely on improved Pubsub mechanisms as provided by libp2p.
+
+### Security: Authentication and authorization for collaborative pinning
+
+In a collaborative pinset scenario, we probably want to have a limited set of peers which are able to modify the shared state and freely connect to API endpoints from any other peers. We call this **trusted peerset**. This implies that we need to find ways to:
+
+* Limit modifications of the state to the  **trusted peerset**, being the rest of cluster peers just "followers".
+* Limit the internal RPC API to the **trusted peerset**
+
+We can address the first problem by signing state upgrades, allowing any peer to authenticate them (as needed). Libp2p pubsub allows sending signed messages so receiving peers can obtain the public key and verify the signatures.
+
+For the second point, we have to consider the internal RPC API surface. Until now, it is assumed that all cluster peers can contact and use the RPC endpoints of all others. This is however very problematic as it would allow any peer for example to trigger ipfs pin/unpins anywhere. For this reason, we propose **authenticated RPC endpoints**, which will only allow a set of trusted peers to execute any of the RPC calls. This can be added as a feature of libp2p-gorpc, taking advantage of the authentication information provided by libp2p. Note, we will have to do UI adjustments so that non-trusted peers receive appropiate errors when they don't have rights to perform certain operations.
```

Comment:
I see why we might need fine grained permissions for a pair of peer and method.

- Peers which are part of the trusted peerset can call any RPC API to other members of the trusted peerset
- But if a peer is not part of the trusted peerset you still want to let it call some RPC APIs, for example ID()

Reply:
How I see it, we are trying to separate permissions for read and writes.

It seems that untrusted peers can't do much, since only thing they can do is to follow. What are the incentives of having a peer that just follows?

Reply:
> What are the incentives of having a peer that just follows?

The same for running IPFS in general. Helping back up content and so on..